### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by aerth**
- [x] Move global vars into a `GameWorld` struct

- [x] create methods from old functions

- [x] fix conflicts from #15

- [x] Condense requests into `msg.Request`

- [x] Condense updates into `msg.Update`

- [x] Add server message logging

- [x] create `GameWorld.ApplyUpdate(*Update)`

### CODE REVIEW

File: client/main.go [hunk 1]
```diff
@@ -129,27 +149,28 @@ func run(addr, id string) error {
 		pixel.R(0, 64, 64, 128),
 		pixel.R(64, 64, 128, 128),
 	}
-
 	playerSprite := pixel.NewSprite(playerSheet, playerFrames[2])
 
-	animationRate := 10.0 // framerate of player animation
-	elapsed := 0.0
-	fps := 0
+	animationRate := 1.0 // framerate of player animation
```
Comment (aerth) [inline client/main.go]: slow down animation to 1 fps, seems to help with jitter for now (until changing player sprite)

File: client/main.go [hunk 2]
```diff
@@ -229,118 +254,133 @@ func loadPicture(path string) (pixel.Picture, error) {
...
 			continue
 		}
+
 	}
 }
 
-func handlePlayerMoved(moved *shared.PlayerMoved) {
-	setPlayerPosition(moved.ID, moved.NewPosition)
-	reapplySimulations(moved.RequestTime)
+func (g *GameWorld) ApplyUpdate(update *shared.Update) {
```
Comment (aerth) [inline client/main.go]: I scrapped the `switch` so that updates can have more than one update type in the future

File: server/main.go [hunk 3]
```diff
@@ -31,9 +35,17 @@ func main() {
...
 	go func() { errc <- serve(*port, errc) }()
 	go func() { gameLoop(errc) }()
-	select {
-	case err := <-errc:
-		log.Fatal("error somewhere", err)
+	for {
+		select {
+		case err := <-errc:
+
+			if strings.HasPrefix(err.Error(), "fatal") {
```
Comment (aerth) [inline server/main.go]: Only quit server if "fatal" errors ( currently only if cant listen on udp port )

File: server/main.go [hunk 4]
```diff
@@ -70,16 +83,27 @@ func serve(port int, errc chan error) error {
...
 			}
-			http.FileServer(http.Dir(".")).ServeHTTP(w, req)
+
+			http.NotFound(w, req)
 		}),
 	)
-	go func() {
-		log.Printf("File server crashed: %v", http.ListenAndServe(laddr, mux))
-	}()
+	if len(clientChecksums) > 0 {
```
Comment (aerth) [inline server/main.go]: only serve tcp if at least 1 client to serve

File: server/main.go (hunk 3 above)
Comment (ilackarms) [inline server/main.go]: i like it. maybe what we can do here is 

```go
switch err.(type){
    case fatalErr:
       log.Fatal(err)
   default:
       log.Println("err: ", err)
}
```

and then define an error type `fatalErr` which we can return

i just prefer this pattern than having to remember to put the prefix `fatal` onto a string

File: server/main.go [hunk 5]
```diff
@@ -96,14 +120,34 @@ func serve(port int, errc chan error) error {
...
 	if err != nil {
+		if strings.Contains(err.Error(), "timeout") {
+			return nil
+		}
 		return err
 	}
-	if msg.ConnectRequest == nil {
+
+	// check first message is ConnectRequest
+	if msg.Request == nil || msg.Request.ConnectRequest == nil {
```
Comment (ilackarms) [inline server/main.go:131]: unneceessary nil check (if err == nil, msg is not nil)

File: server/main.go [hunk 6]
```diff
@@ -70,16 +83,27 @@ func serve(port int, errc chan error) error {
 					w.WriteHeader(http.StatusNoContent)
 					return
 				}
+				if req.URL.Path == "/"+client {
+					clientfile, err := os.Open(client)
```
Comment (ilackarms) [inline server/main.go]: let's use `http.ServeFile` here which will handle for us all the edge cases (proper headers, handling errors, etc.)

File: server/main.go (hunk 4 above)
Comment (ilackarms) [inline server/main.go:97]: :+1:

File: server/main.go [hunk 7]
```diff
@@ -96,14 +120,34 @@ func serve(port int, errc chan error) error {
...
+	// check first message is ConnectRequest
+	if msg.Request == nil || msg.Request.ConnectRequest == nil {
 		return errors.New("expected first message to be ConnectRequest", nil)
 	}
-	id := msg.ConnectRequest.ID
+
+	// get ID
+	id := msg.Request.ConnectRequest.ID
+
+	// check if in use
```
Comment (ilackarms) [inline server/main.go:138]: uncomment or delete? this is a good check to have

File: server/main.go [hunk 8]
```diff
@@ -96,14 +120,34 @@ func serve(port int, errc chan error) error {
...
+	/*	if _, taken := players[id]; taken {
+		return fmt.Errorf("Player ID %q in use", id)
+	} */
+
+	// echo back connect message
```
Comment (ilackarms) [inline server/main.go:143]: why do we echo back?

File: client/main.go [hunk 9]
```diff
@@ -68,52 +78,62 @@ func Run(addr, id string) func() {
...
+		Request: &shared.Request{
+			ConnectRequest: connectionRequest,
+		}}, conn); err != nil {
 		return err
 	}
-	go func() { handleConnection(conn) }()
-	lock.Lock()
-	players[id] = &shared.ClientPlayer{
+
+	msg, err := shared.GetMessage(conn)
```
Comment (ilackarms) [inline client/main.go:108]: what's the purpose of this?

File: server/main.go [hunk 10]
```diff
@@ -151,9 +202,12 @@ func handlePlayer(id string) {
 			})
 			continue
 		}
-		player.QueueLock.Lock()
-		player.RequestQueue = append(player.RequestQueue, msg)
-		player.QueueLock.Unlock()
+		log.Printf("%s %q", msg, id)
+		if msg.Request != nil {
```
Comment (ilackarms) [inline server/main.go:204]: msg should never be nil

File: server/main.go [hunk 11]
```diff
@@ -171,7 +225,7 @@ func gameLoop(errc chan error) {
 		if err := tick(); err != nil {
 			log.Printf("ERROR IN TICK: %v", err)
 			//TODO: handle tick errors
-			//errc <- err
+			errc <- err
```
Comment (ilackarms) [inline server/main.go]: nice
we can remove the TODO here

File: server/main.go [hunk 12]
```diff
@@ -180,11 +234,13 @@ func tick() error {
 	for id, player := range players {
 		player.QueueLock.Lock()
 		for _, msg := range player.RequestQueue {
-			switch {
-			case msg.MoveRequest != nil:
-				handleMoveRequest(id, msg.MoveRequest)
-			case msg.SpeakRequest != nil:
-				handleSpeakRequest(id, msg.SpeakRequest)
+			if msg.Request != nil {
```
Comment (ilackarms) [inline server/main.go:234]: remove

File: server/main.go [hunk 13]
```diff
@@ -256,28 +312,40 @@ func sendWorldState(id string) error {
...
 	if !ok {
+		log.Print(errors.New(fmt.Sprintf("Client disconnected: (cant find ID: %s)", id), nil))
+		playersLock.Lock()
+		delete(players, id)
+		playersLock.Unlock()
+		queueUpdate(&update{
+			notifyPlayerDisconnected: &notifyPlayerDisconnected{
+				id: id,
+			},
+		})
```
Comment (ilackarms) [inline server/main.go]: i don't understand the purpose of this block. if `ok == false`, then `delete(players, id)` is noop

also, unless there is a bug somewhere in the program, `ok` should never `== true` if the player is still connected. if a player disconnects, we should handle the disconnect when it occurs (queue the update notify player disconnected *then*, not wait til this loop to catch it)

i'd be in favor of removing this whole block

**Review by ilackarms — CHANGES_REQUESTED** [review]
see comments

File: client/main.go (hunk 9 above)
Comment (aerth) [reply client/main.go:108]: to confirm connection to server before drawing window

**Review by ilackarms — APPROVED** [review]

**Pull request merged**